src/driver/amdxdna/amdxdna_gem.c

```diff
@@ -213,6 +258,29 @@ amdxdna_drm_create_dev_heap(struct drm_device *dev,
 	return ERR_PTR(ret);
 }
 
+int amdxdna_gem_resv_heap_mem(struct amdxdna_gem_obj *heap, size_t size,
+			      struct drm_mm_node *node, struct amdxdna_mem *mem)
+{
+	struct amdxdna_client *client = heap->client;
+	int ret;
+
+	mutex_lock(&client->mm_lock);
+	ret = amdxdna_gem_insert_node_nolock(heap, size, true, node, mem);
+	mutex_unlock(&client->mm_lock);
+
+	return ret;
+}
+
+void amdxdna_gem_unresv_heap_mem(struct amdxdna_gem_obj *heap, struct drm_mm_node *node,
+				 struct amdxdna_mem *mem)
+{
+	struct amdxdna_client *client = heap->client;
+
+	mutex_lock(&client->mm_lock);
+	amdxdna_gem_remove_node_nolock(node, mem);
+	mutex_unlock(&client->mm_lock);
+}
+
```

Comment:
Why do we need these helpers? Can't we just alloc dev bo for each and every single one of the command slot?

Reply:
We don't really need to allocate drm_gem_object and allocate handle for the reserved memory for command list. Right?
The hwctx will hold ref of the dev heap. It is safe to insert node when hwctx is alive. Thus, I choice this implementation.

Actually, you remind me that I forgot to change amdxdna_drm_alloc_dev_bo() to call amdxdna_gem_insert_node_nolock().

Reply:
> We don't really need to allocate drm_gem_object and allocate handle for the reserved memory for command list. Right? The hwctx will hold ref of the dev heap. It is safe to insert node when hwctx is alive. Thus, I choice this implementation.
> 
> Actually, you remind me that I forgot to change amdxdna_drm_alloc_dev_bo() to call amdxdna_gem_insert_node_nolock().

Would it hurt if we just allocate normal device BO (with drm_gem_object and handle), even though we don't have to?

Reply:
fixed